### PR TITLE
LMN-26 add progress bar component

### PR DESCRIPTION
**Pull request by jonnycaley**
Adds progress bar as per designs [here](https://www.figma.com/file/irZ3YBx8vOm16ICkAr7mB3/Backpack?type=design&node-id=8192-8826&mode=design&t=NpsTDXDDpCd3E0Gv-0)

https://github.com/Skyscanner/backpack-android/assets/26006896/e7959783-48e8-4c80-a87c-2eef1a6fc5f7

### CODE REVIEW

File: .idea/inspectionProfiles/ktlint.xml [hunk 1]
```diff
@@ -8,13 +8,15 @@
     </inspection_tool>
     <inspection_tool class="PreviewApiLevelMustBeValid" enabled="true" level="ERROR" enabled_by_default="true">
       <option name="composableFile" value="true" />
+      <option name="previewFile" value="true" />
```
Comment (bvitaliyg) [inline .idea/inspectionProfiles/ktlint.xml:11]: Are these changes intentional?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    max: Int = 1,
+    stepped: Boolean = false,
+) {
+    val height = if (size == BpkProgressBarSize.Small) 8.dp else 16.dp
+
+    val progress = value
+        .coerceIn(min.toFloat(), max.toFloat())
+        .div(max.toFloat())
+
+    val animatedProgress by animateFloatAsState(
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:64]: Noice!

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                    height = height,
+                    dividers = dividers,
+                )
+            }
+        }
+    }
+}
+
+@Composable
+private fun BuildDividers(
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:95]: I think it's the best to do with onDraw modifier

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                        .background(BpkTheme.colors.surfaceDefault),
+                )
+            }
+        },
+    )
+}
+
+@BpkPreviews
+@Composable
+internal fun PreviewBpkProgressBar() {
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:119]: We usually create the previews inside the app module, and then re-use it for tests and stories

**Review by bvitaliyg — COMMENTED** [review]
Some small comments

File: .idea/inspectionProfiles/ktlint.xml (hunk 1 above)
Comment (jonnycaley) [reply .idea/inspectionProfiles/ktlint.xml:11]: Nah, it was added automatically by IDE

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 4 above)
Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt]: Ahh, I understand now. Removed 🧹

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+enum class BpkProgressBarSize {
+    Small,
+    Large,
+}
+
+@Composable
+fun BpkProgressBar(
+    value: Float,
+    // accessibilityLabel: String, <- not sure if we need this, in its current state talkback reads ("zero percent progress bar")
```
Comment (jonnycaley) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:50]: Not sure if we need this

File: app/src/main/java/net/skyscanner/backpack/demo/compose/ProgressBarStory.kt [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+@Composable
+@ProgressBarComponent
+@ComposeStory
+fun ProgressBarStory(modifier: Modifier = Modifier) {
+
+    Column(modifier = modifier.padding(BpkSpacing.Lg)) {
+
+        ProgressBarSample(
+            title = "Small",
```
Comment (marianeum) [inline app/src/main/java/net/skyscanner/backpack/demo/compose/ProgressBarStory.kt:49]: nit: can you use string resources please?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 5 above)
Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:50]: One thing you could try is adding the `progressSemantics` modifier on the progress indicator and setting the range and value as we get them from here, rather than the transformed version. that way we should get the correct range announced and we're all good

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    modifier: Modifier = Modifier,
+    size: BpkProgressBarSize = BpkProgressBarSize.Small,
+    min: Int = 0,
+    max: Int = 1,
+    stepped: Boolean = false,
+) {
+    val height = if (size == BpkProgressBarSize.Small) 8.dp else 16.dp
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:57]: can you use the `BpkSpacing` values please?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        modifier = Modifier
+            .height(height)
+            .fillMaxWidth(),
+        horizontalArrangement = Arrangement.SpaceEvenly,
+        content = {
+            repeat(dividers) {
+                Box(
+                    modifier = Modifier
+                        .fillMaxHeight()
+                        .width(2.dp)
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:108]: also here `BpkBorderSize.Lg`

File: backpack-lint/src/main/java/net/skyscanner/backpack/lint/check/BpkComposeComponentUsageDetector.kt [hunk 9]
```diff
@@ -83,6 +83,14 @@ class BpkComposeComponentUsageDetector : Detector(), SourceCodeScanner {
                 "androidx.compose.foundation.text.BasicTextField",
             ),
         ),
+        PROGRESS_BAR(
```
Comment (marianeum) [inline backpack-lint/src/main/java/net/skyscanner/backpack/lint/check/BpkComposeComponentUsageDetector.kt:86]: as we no longer expose material we only define this if there's non-material alternatives

**Review by marianeum — COMMENTED** [review]
great first contribution! just a few minor comments

File: app/src/main/java/net/skyscanner/backpack/demo/compose/ProgressBarStory.kt (hunk 6 above)
Comment (jonnycaley) [reply app/src/main/java/net/skyscanner/backpack/demo/compose/ProgressBarStory.kt:49]: Even for story? Why?

Comment (marianeum) [reply app/src/main/java/net/skyscanner/backpack/demo/compose/ProgressBarStory.kt:49]: we're still trying to follow general standards/best practices here

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 5 above)
Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt]: Good idea, done

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 7 above)
Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt]: Done

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 8 above)
Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt]: Done

File: backpack-lint/src/main/java/net/skyscanner/backpack/lint/check/BpkComposeComponentUsageDetector.kt (hunk 9 above)
Comment (jonnycaley) [reply backpack-lint/src/main/java/net/skyscanner/backpack/lint/check/BpkComposeComponentUsageDetector.kt]: Ahh, i did wonder. Done

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+        val numberOfSteps = max
+
+        LinearProgressIndicator(
+            progress = animatedProgress,
+            modifier = Modifier
+                .height(height)
+                .fillMaxWidth()
+                .clip(CircleShape)
+                .progressSemantics(
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: can you also include the values range? that should match the min/max values (that's mainly why they exist in the api)

Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: I left it out because we convert `progress` to a float between 0..1 and coerce it [here](https://github.com/Skyscanner/backpack-android/pull/1658/files/61eceb48942db0db87039c2313242bfb98ce640d#diff-4f9bd9b6cdd30c54e45a5dabb963a30753c5d7bb8a1547788135f320933c02ffR60-R62). As the default `valueRange` parameter is 0..1 we should be able to use that.

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: yeah that's what i meant in my original comment with setting the raw value and the range - that way we'll announce the actual value within the min/max range

Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: Would that be as useful for someone using the screen reader? For example, in the following it will say "0% progress" when it's loaded 50% of the progress which gives off the wrong impression of how much of the data is loaded.

Screenshot 2023-07-21 at 09 21 46

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: I asked @olliecurtis for help, so he will probably come back to you as I'm also not 100% sure what the right approach is here

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: Ok I had a chat with @olliecurtis. We think we can remove min from the API/component as it doesn't make a lot of sense. We should still set the value & valueRange based on max here, but min doesn't make a lot of sense and it should always go from 0 (as it does visually).

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 3 above)
Comment (bvitaliyg) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:102]: It just is what it was designed for. It's also faster in terms of rendering

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt (hunk 10 above)
Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt:79]: I agree. If we're going to change the API, I think it's worth making the following changes:

### Remove `max` and state the value will always be clamped between 0..100

This removes the overhead of "If a parse max in I need to parse a value in between 0..max and not 0..1.", which will be required if users want to add steps to the component.

### Change `value: Float` to `percentage: Int`

When the progress is an integer (this is true for [flights](StaggeredProgressBarView) and hotels - we receive `"completionPercentage": Int` in the DTO), users have to convert to a `Float`, which looks like it will be most of the time.

### Change `stepped (Boolean)` to `numberOfSteps (Int)` and set it default to 1

Asking "How many steps are there to the progress?" with a single variable is easier to understand than "The number of steps will be max - 1, and by default max is 1 so you will not see any steps even if you set `stepped == true` if you do not set max != 1".

So will end up like this:

```
@Composable
fun BpkProgressBar(
    percentage: Int, // value between 0..100
    modifier: Modifier = Modifier,
    size: BpkProgressBarSize = BpkProgressBarSize.Small,
    numberOfSteps: Int = 1, // 1..100
)
```

Wdyt?

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt]: Had another chat with @olliecurtis and it still makes sense to support the max value for both the stepped and non-stepped version as you may have some extra information about the maximum value. for example of web we get back x providers out of y and the progress bar indicates that. 
i think having `numberOfSteps` combined with `percentage` can be equally confusing as you need to do mapping between the percentage and step count. also you may have 55.5% for example and an int may not be enough.

so to keep things in sync across platforms we will need to maintain this (min will be removed on web too).

Comment (jonnycaley) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/progressbar/BpkProgressBar.kt]: Fair enough, done
